packages/block-editor/src/components/list-view/branch.js

```diff
@@ -29,57 +29,42 @@ import useBlockDisplayInformation from '../use-block-display-information';
  * @param {Object}  expandedState       state that notes which branches are collapsed
  * @param {Array}   draggedClientIds    a list of dragged client ids
  * @param {boolean} isExpandedByDefault flag to determine the default fallback expanded state.
+ * @param {number?} sum                 current sum for recursion
  * @return {number} block count
  */
 function countBlocks(
 	block,
 	expandedState,
 	draggedClientIds,
-	isExpandedByDefault
+	isExpandedByDefault,
+	sum = 0
 ) {
 	const isDragged = draggedClientIds?.includes( block.clientId );
 	if ( isDragged ) {
 		return 0;
 	}
+
 	const isExpanded = expandedState[ block.clientId ] ?? isExpandedByDefault;
+	if ( ! isExpanded || ! block.innerBlocks.length ) {
+		return 1;
+	}
 
-	if ( isExpanded ) {
-		return (
-			1 +
-			block.innerBlocks.reduce(
-				countReducer(
+	return (
+		sum +
+		block.innerBlocks.reduce(
+			( innerSum, innerBlock ) =>
+				innerSum +
+				countBlocks(
+					innerBlock,
 					expandedState,
 					draggedClientIds,
-					isExpandedByDefault
+					isExpandedByDefault,
+					sum
 				),
-				0
-			)
-		);
-	}
-	return 1;
+			0
+		)
+	);
 }
-const countReducer =
-	( expandedState, draggedClientIds, isExpandedByDefault ) =>
-	( count, block ) => {
-		const isDragged = draggedClientIds?.includes( block.clientId );
-		if ( isDragged ) {
-			return count;
-		}
-		const isExpanded =
-			expandedState[ block.clientId ] ?? isExpandedByDefault;
-		if ( isExpanded && block.innerBlocks.length > 0 ) {
-			return (
-				count +
-				countBlocks(
-					block,
-					expandedState,
-					draggedClientIds,
-					isExpandedByDefault
-				)
-			);
-		}
-		return count + 1;
-	};
```

Comment:
I was just trying to figure out what this was doing by incrementally simplifying it. The code can go into another PR like the other little refactors I've been doing as I've been reading through the code.